core/src/main/scala/kafka/log/LogValidator.scala

```diff
@@ -210,6 +141,87 @@ private[kafka] object LogValidator {
+                                                 compactedTopic: Boolean = false,
+                                                 messageFormatVersion: Byte = Record.CURRENT_MAGIC_VALUE,
+                                                 messageTimestampType: TimestampType,
+                                                 messageTimestampDiffMaxMs: Long): ValidationAndOffsetAssignResult = {
+    // No in place assignment situation 1 and 2
+    var inPlaceAssignment = sourceCodec == targetCodec && messageFormatVersion > Record.MAGIC_VALUE_V0
+
+    var maxTimestamp = Record.NO_TIMESTAMP
+    val expectedInnerOffset = new LongRef(0)
+    val validatedRecords = new mutable.ArrayBuffer[Record]
+
+    records.deepEntries(true).asScala.foreach { logEntry =>
+      val record = logEntry.record
+      validateKey(record, compactedTopic)
+
+      if (record.magic > Record.MAGIC_VALUE_V0 && messageFormatVersion > Record.MAGIC_VALUE_V0) {
+        // Validate the timestamp
+        validateTimestamp(record, now, messageTimestampType, messageTimestampDiffMaxMs)
+        // Check if we need to overwrite offset, no in place assignment situation 3
+        if (logEntry.offset != expectedInnerOffset.getAndIncrement())
+          inPlaceAssignment = false
+        if (record.timestamp > maxTimestamp)
+          maxTimestamp = record.timestamp
+      }
+
+      if (sourceCodec != NoCompressionCodec && logEntry.isCompressed)
+        throw new InvalidMessageException("Compressed outer record should not have an inner record with a " +
+          s"compression attribute set: $record")
+
+      // No in place assignment situation 4
+      if (record.magic != messageFormatVersion)
+        inPlaceAssignment = false
+
+      validatedRecords += record.convert(messageFormatVersion)
+    }
+
+    if (!inPlaceAssignment) {
+      val entries = validatedRecords.map(record => LogEntry.create(offsetCounter.getAndIncrement(), record))
+      val builder = MemoryRecords.builderWithEntries(messageTimestampType, CompressionType.forId(targetCodec.codec),
+        now, entries.asJava)
+      val updatedRecords = builder.build()
+      val info = builder.info
+      ValidationAndOffsetAssignResult(
+        validatedRecords = updatedRecords,
+        maxTimestamp = info.maxTimestamp,
+        shallowOffsetOfMaxTimestamp = info.shallowOffsetOfMaxTimestamp,
+        messageSizeMaybeChanged = true)
+    } else {
+      // ensure the inner messages are valid
+      validatedRecords.foreach(_.ensureValid)
+
+      // we can update the wrapper message only and write the compressed payload as is
+      val entry = records.shallowEntries.iterator.next()
+      val offset = offsetCounter.addAndGet(validatedRecords.size) - 1
+      entry.setOffset(offset)
+
+      val shallowTimestamp = if (messageTimestampType == TimestampType.LOG_APPEND_TIME) now else maxTimestamp
+      if (messageTimestampType == TimestampType.LOG_APPEND_TIME)
+        entry.setLogAppendTime(shallowTimestamp)
```

Comment:
I guess one thing we could consider doing in another followup is replacing `setLogAppendTime` and `setCreateTime` with a single method accepting `TimestampType` as an argument.